htlcswitch/switch.go

```diff
@@ -0,0 +1,399 @@
+			diffSatSent := overallSatSent - prevSatSent
+			diffSatRecv := overallSatRecv - prevSatRecv
+
+			log.Infof("Sent %v satoshis, received %v satoshi in "+
+				"the last 10 seconds (%v tx/sec)",
+				diffSatSent.ToUnit(btcutil.AmountSatoshi),
+				diffSatRecv.ToUnit(btcutil.AmountSatoshi),
+				float64(diffNumUpdates)/10)
+
+			prevNumUpdates = overallNumUpdates
+			prevSatSent = overallSatSent
+			prevSatRecv = overallSatRecv
+
+		case <-s.service.Quit:
+			return
+		}
+	}
+}
+
+// Start starts all helper goroutines required for the operation of the switch.
+func (s *HTLCSwitch) Start() error {
+	log.Infof("HTLC Switch starting")
+	if err := s.service.Start(); err != nil {
+		return err
+	}
+
+	s.service.Go(s.startHandling)
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+func (s *HTLCSwitch) Stop() error {
+	log.Infof("HTLC Switch shutting down")
+	if err := s.service.Stop(nil); err != nil {
+		return err
+	}
+
+	for _, manager := range s.managers {
+		s.remove(manager)
+	}
+
+	return nil
+}
+
+// Add is used to add and start the newly created HTLC manager ans start use it
+// to propagate HTLCs.
+func (s *HTLCSwitch) Add(manager HTLCManager) error {
+	if err := manager.Start(); err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	s.managers[*manager.ID()] = manager
+	s.mutex.Unlock()
+
+	log.Infof("Added htlc manager for ChannelPoint(%v), bandwidth=%v",
+		manager.ID(), manager.Bandwidth())
+	return nil
+}
+
+// Remove is used to remove and stop the htlc manager by channel point of the
+// channel which htlc manager is managing.
+func (s *HTLCSwitch) RemoveByChan(chanPoint *wire.OutPoint) error {
+	manager, err := s.Get(chanPoint)
+	if err != nil {
+		return err
+	}
+
+	return s.remove(manager)
+}
+
+// Get returns the htlc manager which corresponds to the channel which he is
+// managing.
+func (s *HTLCSwitch) Get(chanPoint *wire.OutPoint) (HTLCManager, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	manager, ok := s.managers[*chanPoint]
+	if !ok {
+		return nil, ErrHTLCManagerNotFound
+	}
+
+	return manager, nil
+}
+
+// RemoveById removes all HTLC managers which correspond to destination id.
+func (s *HTLCSwitch) RemoveById(id *routing.HopID) error {
+	managers, err := s.getManagersByDest(id)
+	if err != nil {
+		return err
+	}
+
+	for _, manager := range managers {
+		if err := s.remove(manager); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// remove is helper function which removes and stops HTLC manager.
+func (s *HTLCSwitch) remove(manager HTLCManager) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	chanPoint := manager.ID()
+	delete(s.managers, *chanPoint)
+	manager.Stop()
+
+	log.Infof("Remove htlc manager for ChannelPoint(%v)", manager.ID())
+	return nil
+}
+
+// getManagersByDest is helper function which returns the htlc managers by hop
+// destination id.
+func (s *HTLCSwitch) getManagersByDest(destination *routing.HopID) ([]HTLCManager,
```

Comment:
Usage of an additional index for interfaces (node pubkeys) would eliminate the need for this function and would also be a performance improvement as it would allows us to remove much of the mutex usage.